internal/pkg/providers/azure/azure.go

```diff
@@ -0,0 +1,238 @@
+package azure
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/profiles/latest/network/mgmt/network"
+	"github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2019-05-01/resources"
+
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/mitchellh/mapstructure"
+	"github.com/sirupsen/logrus"
+	"github.com/stakater/Whitelister/internal/pkg/config"
+	clientset "k8s.io/client-go/kubernetes"
+
+	"github.com/stakater/Whitelister/internal/pkg/utils"
+)
+
+// Azure provider class implementing the Provider interface
+type Azure struct {
+	ClientSet                 clientset.Interface
```

Comment:
Maybe keep clients in this structure instead of credentials? e.g `securityRulesClient`, `securityGroupClient` etc. Then we don't need to create new clients on each reconcile loop.